File: 15_Semantic_Rules/ets_cases/15.7_Type_Inference/cross_lang_verify/SEM_15_07_Inference.swift
/**
 * Swift cross-language verification for 15.7 Type Inference
 *
 * Key finding: Swift has stronger type inference than ArkTS/Java.
 * - Return type inference supported
 * - Generic type inference
 */
import Foundation

var pass = 0, fail = 0
func check(_ cond: Bool, _ msg: String) {
    if cond { pass += 1; print("  PASS \(msg)") }
    else { fail += 1; print("  FAIL \(msg)") }
}

// Return type inference: Swift supports it
func add(_ a: Int, _ b: Int) -> Int { return a + b }

// Generic type inference
func identity<T>(_ x: T) -> T { return x }

// Local type inference
let x = 42
check(x == 42, "local type inference (Swift: same)")

// Return type
check(add(1, 2) == 3, "return type (Swift: supports type inference)")

// Generic type inference
let r: Int = identity(42)
check(r == 42, "generic type inference (Swift: same)")

// Constant expression
let c = 1 + 2
check(c == 3, "constant expression (Swift: same)")

print("\n=== Summary ===")
print("Pass: \(pass)  Fail: \(fail)")
if fail > 0 { exit(1) }
print("verified")

File: 15_Semantic_Rules/ets_cases/15.7_Type_Inference/cross_lang_verify/SEM_15_07_Inference.java
/**
 * Java cross-language verification for 15.7 Type Inference
 *
 * Key findings: Similar across all three, but Swift has stronger inference.
 * - Local type inference from initializer
 * - Return type inference
 * - Generic type inference
 */
class SEM_15_07_Inference {
    static int pass = 0, fail = 0;
    static void check(boolean cond, String msg) {
        if (cond) { pass++; System.out.println("  PASS " + msg); }
        else { fail++; System.out.println("  FAIL " + msg); }
    }

    // Return type inference: Java requires explicit return type
    static int add(int a, int b) { return a + b; }

    // Generic type inference
    static <T> T identity(T x) { return x; }

    public static void main(String[] args) {
        // Local type inference (explicit type, Java 8 compatible)
        int x = 42;
        check(x == 42, "local type inference (Java: explicit type)");

        // Return type: explicit in Java
        check(add(1, 2) == 3, "return type (Java: explicit, unlike ArkTS)");

        // Generic type inference
        Integer r = identity(42);
        check(r == 42, "generic type inference (Java: same)");

        // Constant expression type inference
        final int c = 1 + 2;
        check(c == 3, "constant expression (Java: same)");

        System.out.println("\n=== Summary ===");
        System.out.println("Pass: " + pass + "  Fail: " + fail);
        if (fail > 0) System.exit(1);
        System.out.println("verified");
    }
}
